Improvements for guest dump

mhartmay opened the pull request.

mhartmay commented:
Note: I'm not totally sure why the trick with the temporary file was used... Permissions? To delete the file if there was an error?

arighi commented:
> Note: I'm not totally sure why the trick with the temporary file was used... Permissions? To delete the file if there was an error?
> 
> 2. And what is the benefit of creating a temporary file that is automatically removed (if no one access it)? To test if guest dumping works? (if `args.dump is None`)

So, the temporary file is convenient if you want to dump to the same file multiple times. For example, now I get this when I try the second time:
```
$ vng --dump /tmp/vmcore.img
Dump failed:
{"error": {"class": "GenericError", "desc": "Could not create '/tmp/vmcore.img': Permission denied"}}
```

That's because the dump file is created as with permissions 0400. Maybe we can just a chmod 600 after the dump has been created, if we don't want to do the temp file trick?

mhartmay commented:
> > So our expectation is that an existing file is getting (silently) overwritten?
> 
> Personally, I think it makes sense to fail in this case, but then with a better error message. (but I guess it is also fine to overwrite if we prefer).

Right now, it just overwrites the file. Not sure if we're allowed to change this behavior.

> 
> > How should we handle the case that an dump is requested, but no output file is selected? I guess that's not even possible with our current CLI?
> 
> For the moment, `--dump` has to be used with an argument, no? It makes sense to keep this like that, no?

Yep that's also how the new code works. The old code handled the case where the argument was not specified. I've removed this and replaced it with an assert statement: `assert args.dump is not None`.

matttbe commented:
> > > > How should we handle the case that an dump is requested, but no output file is selected? I guess that's not even possible with our current CLI?
> > > 
> > > 
> > > For the moment, `--dump` has to be used with an argument, no? It makes sense to keep this like that, no?
> > 
> > 
> > Yep that's also how the new code works. The old code handled the case where the argument was not specified. I've removed this and replaced it with an assert statement: `assert args.dump is not None`.
> 
> I see. I don't even think this `assert` is needed, we don't do that elsewhere.

From what I see, the old code was doing `shutil.move(tmp.name, dump_file)` with `dump_file == None`, so not fully handling this case :)

mhartmay commented on virtme_ng/run.py:
```diff
@@ -1397,38 +1397,73 @@ def dump(self, args):
+        if json.loads(data) != {"return": {}}:
+            sys.stderr.write(f"Dump failed:\n{data}")
+            sys.exit(1)
+        assert args.dump is not None
```
Okay, will remove the assertion then and leave everything else as is. Fine with you?

matttbe replied:
> Okay, will remove the assertion then and leave everything else as is. Fine with you?

Yes, fine with me. (and just to be clear, I'm not against assertion, but also good to uniform the code)

mhartmay commented:
> > > > So our expectation is that an existing file is getting (silently) overwritten?
> > > 
> > > 
> > > Personally, I think it makes sense to fail in this case, but then with a better error message. (but I guess it is also fine to overwrite if we prefer).
> > 
> > 
> > Right now, it just overwrites the file. Not sure if we're allowed to change this behavior.
> 
> I think that can be seen as a fix. That's a specific command, it should be fine.

@arighi What's your opinion about that?

[...snip]

arighi commented:
> > > > > So our expectation is that an existing file is getting (silently) overwritten?
> > > > 
> > > > 
> > > > Personally, I think it makes sense to fail in this case, but then with a better error message. (but I guess it is also fine to overwrite if we prefer).
> > > 
> > > 
> > > Right now, it just overwrites the file. Not sure if we're allowed to change this behavior.
> > 
> > 
> > I think that can be seen as a fix. That's a specific command, it should be fine.
> 
> @arighi What's your opinion on that?
> 
> [...snip]

Sorry I missed this one. After re-reading the whole discussion I also think it's safer to not overwrite the file and this can actually be seen as a fix.
